File: backend/app/services/gmail_oauth.py
"""One-time Gmail OAuth setup and Gmail API access-token refresh helpers.

The public application never asks ordinary members for Gmail access. An
authenticated admin authorizes the single sender mailbox once; the resulting
refresh token is kept outside Git and reused by the password-reset mailer.
"""
from __future__ import annotations

import base64
from dataclasses import dataclass
from datetime import datetime, timezone
import hashlib
import json
import logging
import os
from pathlib import Path
import secrets
from threading import Lock
import time
from typing import Any
from urllib.error import HTTPError, URLError
from urllib.parse import urlencode
from urllib.request import Request, urlopen

from ..core.config import get_settings


logger = logging.getLogger("recsys.gmail_oauth")
GMAIL_SEND_SCOPE = "https://www.googleapis.com/auth/gmail.send"
_AUTH_URI = "https://accounts.google.com/o/oauth2/auth"
_TOKEN_URI = "https://oauth2.googleapis.com/token"
_PENDING_TTL_SECONDS = 10 * 60
_pending: dict[str, tuple[float, str]] = {}
_pending_lock = Lock()
_access_cache: tuple[str, float] | None = None
_access_lock = Lock()


class GmailOAuthError(RuntimeError):
    """Safe, non-secret error raised by the Gmail OAuth integration."""


@dataclass(frozen=True)
class OAuthClient:
    client_id: str
    client_secret: str
    auth_uri: str
    token_uri: str
    redirect_uri: str


def _read_json(path: Path) -> dict[str, Any]:
    try:
        payload = json.loads(path.read_text(encoding="utf-8"))
    except (OSError, ValueError, TypeError) as exc:
        raise GmailOAuthError("Google OAuth credential file is unreadable") from exc
    if not isinstance(payload, dict):
        raise GmailOAuthError("Google OAuth credential file is invalid")
    return payload


def load_client() -> OAuthClient:
    settings = get_settings()
    client_id = str(getattr(settings, "gmail_oauth_client_id", "") or "").strip()
    client_secret = str(getattr(settings, "gmail_oauth_client_secret", "") or "").strip()
    auth_uri = _AUTH_URI
    token_uri = _TOKEN_URI

    client_file = getattr(settings, "gmail_oauth_client_file", None)
    if (not client_id or not client_secret) and client_file:
        payload = _read_json(Path(client_file))
        web = payload.get("web")
        if not isinstance(web, dict):
            raise GmailOAuthError("OAuth credential must be a Web application client")
        client_id = str(web.get("client_id") or "").strip()
        client_secret = str(web.get("client_secret") or "").strip()
        auth_uri = str(web.get("auth_uri") or _AUTH_URI).strip()
        token_uri = str(web.get("token_uri") or _TOKEN_URI).strip()

    redirect_uri = str(
        getattr(settings, "gmail_oauth_redirect_uri", "")
        or "http://localhost:8001/auth/google/callback"
    ).strip()
    if not client_id or not client_secret:
        raise GmailOAuthError("Google OAuth client is not configured")
    return OAuthClient(client_id, client_secret, auth_uri, token_uri, redirect_uri)


def client_configured() -> bool:
    try:
        load_client()
        return True
    except GmailOAuthError:
        return False


def _refresh_token() -> str:
    settings = get_settings()
    env_token = str(getattr(settings, "gmail_oauth_refresh_token", "") or "").strip()
    if env_token:
        return env_token
    token_file = Path(getattr(settings, "gmail_oauth_token_file"))
    if not token_file.is_file():
        return ""
    try:
        return str(_read_json(token_file).get("refresh_token") or "").strip()
    except GmailOAuthError:
        return ""


def authorized() -> bool:
    return client_configured() and bool(_refresh_token())


def start_authorization() -> str:
    """Return a Google consent URL and remember one short-lived PKCE state."""
    client = load_client()
    state = secrets.token_urlsafe(32)
    verifier = secrets.token_urlsafe(64)
    challenge = base64.urlsafe_b64encode(
        hashlib.sha256(verifier.encode("ascii")).digest()
    ).rstrip(b"=").decode("ascii")
    now = time.time()
    with _pending_lock:
        for key, (created_at, _) in list(_pending.items()):
            if now - created_at > _PENDING_TTL_SECONDS:
                _pending.pop(key, None)
        _pending[state] = (now, verifier)
    return f"{client.auth_uri}?{urlencode({
        'client_id': client.client_id,
        'redirect_uri': client.redirect_uri,
        'response_type': 'code',
        'scope': GMAIL_SEND_SCOPE,
        'access_type': 'offline',
        'prompt': 'consent',
        'include_granted_scopes': 'true',
        'state': state,
        'code_challenge': challenge,
        'code_challenge_method': 'S256',
    })}"


def complete_authorization(code: str, state: str) -> None:
    """Exchange a one-time authorization code and persist its refresh token."""
    with _pending_lock:
        pending = _pending.pop(state, None)
    if pending is None or time.time() - pending[0] > _PENDING_TTL_SECONDS:
        raise GmailOAuthError("OAuth state is invalid or expired")
    verifier = pending[1]
    client = load_client()
    payload = _post_form(
        client.token_uri,
        {
            "code": code,
            "client_id": client.client_id,
            "client_secret": client.client_secret,
            "redirect_uri": client.redirect_uri,
            "grant_type": "authorization_code",
            "code_verifier": verifier,
        },
    )
    refresh_token = str(payload.get("refresh_token") or "").strip()
    if not refresh_token:
        raise GmailOAuthError("Google did not return a refresh token; revoke access and try again")
    settings = get_settings()
    token_file = Path(settings.gmail_oauth_token_file)
    token_file.parent.mkdir(parents=True, exist_ok=True)
    token_file.write_text(
        json.dumps(
            {
                "refresh_token": refresh_token,
                "scope": str(payload.get("scope") or GMAIL_SEND_SCOPE),
                "created_at": datetime.now(timezone.utc).isoformat(),
            },
            ensure_ascii=False,
            indent=2,
        ),
        encoding="utf-8",
    )
    try:
        os.chmod(token_file, 0o600)
    except OSError:
        logger.warning("Could not tighten permissions on Gmail OAuth token file")
    global _access_cache
    _access_cache = None


def get_access_token() -> str:
    """Refresh and cache a short-lived Gmail API access token."""
    global _access_cache
    now = time.time()
    with _access_lock:
        if _access_cache and _access_cache[1] > now + 60:
            return _access_cache[0]
        client = load_client()
        refresh_token = _refresh_token()
        if not refresh_token:
            raise GmailOAuthError("Gmail sender has not been authorized")
        payload = _post_form(
            client.token_uri,
            {
                "client_id": client.client_id,
                "client_secret": client.client_secret,
                "refresh_token": refresh_token,
                "grant_type": "refresh_token",
            },
        )
        access_token = str(payload.get("access_token") or "").strip()
        if not access_token:
            raise GmailOAuthError("Google did not return an access token")
        expires_in = max(120, int(payload.get("expires_in") or 3600))
        _access_cache = (access_token, now + expires_in)
        return access_token


def _post_form(url: str, fields: dict[str, str]) -> dict[str, Any]:
    body = urlencode(fields).encode("utf-8")
    request = Request(
        url,
        data=body,
        headers={"Content-Type": "application/x-www-form-urlencoded"},
        method="POST",
    )
    try:
        with urlopen(request, timeout=20) as response:
            payload = json.loads(response.read().decode("utf-8"))
    except HTTPError as exc:
        # Do not include Google's response body: it can contain sensitive data.
        raise GmailOAuthError(f"Google OAuth request failed ({exc.code})") from exc
    except (URLError, OSError, ValueError) as exc:
        raise GmailOAuthError("Google OAuth request failed") from exc
    if not isinstance(payload, dict):
        raise GmailOAuthError("Google OAuth response is invalid")
    return payload

File: frontend/lib/server/gmail-oauth.ts
// One-time Gmail sender authorization and access-token refresh (admin only).
//
// This client is intentionally separate from member Google Login. It requests
// the narrow ``gmail.send`` scope, and its refresh token is persisted outside
// Git (env var or a private JSON file) for the password-reset mailer.
import { createHash, randomBytes } from "node:crypto";
import { chmod, mkdir, readFile, writeFile } from "node:fs/promises";
import { dirname } from "node:path";
import { postForm } from "@/lib/server/oauth-http";

export const GMAIL_STATE_COOKIE = "thaiperform_gmail_oauth_state";
const GMAIL_SEND_SCOPE = "https://www.googleapis.com/auth/gmail.send";
const DEFAULT_AUTH_URI = "https://accounts.google.com/o/oauth2/auth";
const DEFAULT_TOKEN_URI = "https://oauth2.googleapis.com/token";
const PENDING_TTL_SECONDS = 10 * 60;
const DEFAULT_GMAIL_API_SEND_URI = "https://gmail.googleapis.com/gmail/v1/users/me/messages/send";

export class GmailOAuthError extends Error {}

interface GmailClient {
  clientId: string;
  clientSecret: string;
  authUri: string;
  tokenUri: string;
  redirectUri: string;
}

function envValue(name: string): string {
  return (process.env[name] ?? "").trim();
}

// fallow-ignore-next-line complexity -- Unreadable and malformed credential files must each fail closed.
async function readClientFile(path: string | undefined): Promise<Record<string, unknown> | null> {
  if (!path) return null;
  try {
    const parsed = JSON.parse(await readFile(path, "utf8")) as Record<string, unknown>;
    return parsed && typeof parsed === "object" ? parsed : null;
  } catch {
    return null;
  }
}

async function clientWeb(path: string | undefined): Promise<Record<string, unknown> | null> {
  const parsed = await readClientFile(path);
  if (!parsed) return null;
  const web = parsed.web as Record<string, unknown> | undefined;
  if (!web || typeof web !== "object") {
    throw new GmailOAuthError("OAuth credential must be a Web application client");
  }
  return web;
}

// fallow-ignore-next-line complexity -- Env-var and client-file fallbacks must resolve in one deterministic precedence order.
async function loadGmailClient(): Promise<GmailClient> {
  const web = await clientWeb(envValue("GMAIL_OAUTH_CLIENT_FILE"));
  const clientId = envValue("GMAIL_OAUTH_CLIENT_ID") || String(web?.client_id ?? "").trim();
  const clientSecret = envValue("GMAIL_OAUTH_CLIENT_SECRET") || String(web?.client_secret ?? "").trim();
  const authUri = envValue("GMAIL_OAUTH_AUTH_URI") || String(web?.auth_uri ?? "").trim() || DEFAULT_AUTH_URI;
  const tokenUri = envValue("GMAIL_OAUTH_TOKEN_URI") || String(web?.token_uri ?? "").trim() || DEFAULT_TOKEN_URI;
  const redirectUri = envValue("GMAIL_OAUTH_REDIRECT_URI");
  if (!clientId || !clientSecret || !redirectUri) {
    throw new GmailOAuthError("Google OAuth client is not configured");
  }
  return { clientId, clientSecret, authUri, tokenUri, redirectUri };
}

async function gmailTokenFilePath(): Promise<string> {
  return envValue("GMAIL_OAUTH_TOKEN_FILE") || "";
}

// fallow-ignore-next-line complexity -- Env and token-file sources must each be probed without leaking secrets.
async function refreshToken(): Promise<string> {
  const envToken = envValue("GMAIL_OAUTH_REFRESH_TOKEN");
  if (envToken) return envToken;
  const tokenFile = await gmailTokenFilePath();
  if (!tokenFile) return "";
  try {
    const payload = JSON.parse(await readFile(tokenFile, "utf8")) as { refresh_token?: unknown };
    return String(payload.refresh_token ?? "").trim();
  } catch {
    return "";
  }
}

export async function clientConfigured(): Promise<boolean> {
  try {
    await loadGmailClient();
    return true;
  } catch {
    return false;
  }
}

export async function authorized(): Promise<boolean> {
  return (await clientConfigured()) && Boolean(await refreshToken());
}

/** Build the Google consent URL and remember one short-lived PKCE flow. */
export async function startAuthorization(): Promise<{ authorizationUrl: string; state: string; verifier: string }> {
  const client = await loadGmailClient();
  const state = randomBytes(24).toString("base64url");
  const verifier = randomBytes(48).toString("base64url");
  const challenge = createHash("sha256").update(verifier, "ascii").digest("base64url");
  const params = new URLSearchParams({
    client_id: client.clientId,
    redirect_uri: client.redirectUri,
    response_type: "code",
    scope: GMAIL_SEND_SCOPE,
    access_type: "offline",
    prompt: "consent",
    include_granted_scopes: "true",
    state,
    code_challenge: challenge,
    code_challenge_method: "S256",
  });
  return { authorizationUrl: `${client.authUri}?${params.toString()}`, state, verifier };
}

/** Exchange the one-time code and persist the refresh token for the mailer. */
// fallow-ignore-next-line complexity -- Token exchange, persistence, and cache invalidation are one security boundary.
export async function completeAuthorization(code: string, verifier: string): Promise<void> {
  const client = await loadGmailClient();
  let payload: Record<string, unknown>;
  try {
    payload = await postForm(
      client.tokenUri,
      {
        code,
        client_id: client.clientId,
        client_secret: client.clientSecret,
        redirect_uri: client.redirectUri,
        grant_type: "authorization_code",
        code_verifier: verifier,
      },
      "Google OAuth request failed",
    );
  } catch (error) {
    throw new GmailOAuthError(error instanceof Error ? error.message : "Google OAuth request failed");
  }
  const refreshTokenValue = String(payload.refresh_token ?? "").trim();
  if (!refreshTokenValue) {
    throw new GmailOAuthError("Google did not return a refresh token; revoke access and try again");
  }
  const tokenFile = await gmailTokenFilePath();
  if (tokenFile) {
    await mkdir(dirname(tokenFile), { recursive: true });
    await writeFile(
      tokenFile,
      JSON.stringify(
        {
          refresh_token: refreshTokenValue,
          scope: String(payload.scope ?? GMAIL_SEND_SCOPE),
          created_at: new Date().toISOString(),
        },
        null,
        2,
      ),
      "utf8",
    );
    try {
      await chmod(tokenFile, 0o600);
    } catch {
      // Windows filesystems have no POSIX mode bits; best effort only.
    }
  }
  accessTokenCache = undefined;
}

let accessTokenCache: { token: string; expiresAt: number } | undefined;

/** Refresh and cache a short-lived Gmail API access token. */
// fallow-ignore-next-line complexity -- Cache expiry and refresh failure handling are one bounded token boundary.
async function getAccessToken(): Promise<string> {
  const now = Date.now();
  if (accessTokenCache && accessTokenCache.expiresAt > now + 60_000) {
    return accessTokenCache.token;
  }
  const client = await loadGmailClient();
  const storedRefreshToken = await refreshToken();
  if (!storedRefreshToken) throw new GmailOAuthError("Gmail sender has not been authorized");
  let payload: Record<string, unknown>;
  try {
    payload = await postForm(
      client.tokenUri,
      {
        client_id: client.clientId,
        client_secret: client.clientSecret,
        refresh_token: storedRefreshToken,
        grant_type: "refresh_token",
      },
      "Google OAuth request failed",
    );
  } catch (error) {
    throw new GmailOAuthError(error instanceof Error ? error.message : "Google OAuth request failed");
  }
  const accessToken = String(payload.access_token ?? "").trim();
  if (!accessToken) throw new GmailOAuthError("Google did not return an access token");
  const expiresIn = Math.max(120, Number(payload.expires_in) || 3600);
  accessTokenCache = { token: accessToken, expiresAt: now + expiresIn * 1000 };
  return accessToken;
}

/** Deliver one RFC 2822 message with the narrowly scoped Gmail API. */
export async function sendWithGmailApi(rawMessage: string): Promise<boolean> {
  try {
    const accessToken = await getAccessToken();
    const response = await fetch(envValue("GMAIL_API_SEND_URI") || DEFAULT_GMAIL_API_SEND_URI, {
      method: "POST",
      headers: {
        Authorization: `Bearer ${accessToken}`,
        "Content-Type": "application/json",
      },
      body: JSON.stringify({ raw: Buffer.from(rawMessage, "utf8").toString("base64url") }),
      cache: "no-store",
      signal: AbortSignal.timeout(20_000),
    });
    return response.ok;
  } catch {
    return false;
  }
}

export function pendingTtlSeconds(): number {
  return PENDING_TTL_SECONDS;
}
